File: src/main/java/com/university/repositories/TravelAllowanceRepository.java
package com.university.repositories;

import com.university.entities.City;
import com.university.entities.TravelAllowance;
import com.university.entities.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDate;
import java.util.List;

public interface TravelAllowanceRepository extends JpaRepository<TravelAllowance, Integer> {

    List<TravelAllowance> findTravelAllowancesByUser(User user);

    Integer countByCity(City city);

    @Query("select t from TravelAllowance t where " +
            "(-1=:department OR t.user.department.id=:department) and " +
            "(-1=:position OR t.user.position.id=:position) and " +
            "(0=:year OR YEAR(t.dateOfIssue)=:year) and " +
            "(0=:month OR MONTH(t.dateOfIssue)=:month)")
    List<TravelAllowance> findByFilter(@Param("department") int department,
                                       @Param("position") int position,
                                       @Param("year") int year,
                                       @Param("month") int month);

    List<TravelAllowance> findTravelAllowancesByUserAndDateOfIssueBetween(User user, LocalDate start, LocalDate end);
}

File: src/main/kotlin/com/university/repositories/TravelAllowanceRepository.kt
package com.university.repositories

import com.university.entities.City
import com.university.entities.TravelAllowance
import com.university.entities.User
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.LocalDate

@Repository
interface TravelAllowanceRepository : JpaRepository<TravelAllowance, Int> {
    fun findTravelAllowancesByUser(user: User): List<TravelAllowance>
    fun countByCity(city: City): Int

    @Query("select t from TravelAllowance t where " +
            "(-1=:department OR t.user.department.id=:department) and " +
            "(-1=:position OR t.user.position.id=:position) and " +
            "(0=:year OR YEAR(t.dateOfIssue)=:year) and " +
            "(0=:month OR MONTH(t.dateOfIssue)=:month)")
    fun findByFilter(@Param("department") department: Int,
                     @Param("position") position: Int,
                     @Param("year") year: Int,
                     @Param("month") month: Int): List<TravelAllowance>

    fun findTravelAllowancesByUserAndDateOfIssueBetween(user: User, start: LocalDate, end: LocalDate): List<TravelAllowance>
}
